include/aws/crt/ByteBuf.h

```diff
@@ -0,0 +1,109 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/crt/Outcome.h>
+#include <aws/crt/Types.h>
+
+#include <aws/common/byte_buf.h>
+
+namespace Aws
+{
+    namespace Crt
+    {
+        class ByteCursor
+        {
+          public:
+            ByteCursor() noexcept;
+            ByteCursor(const ByteCursor &cursor) noexcept;
+            explicit ByteCursor(const char *str) noexcept;
+            explicit ByteCursor(const String &str) noexcept;
+            explicit ByteCursor(aws_byte_cursor cursor) noexcept;
+            explicit ByteCursor(const aws_byte_buf &buffer) noexcept;
+            ByteCursor(const uint8_t *array, size_t len) noexcept;
+
+            ByteCursor &operator=(const ByteCursor &cursor) noexcept;
+
+            /**
+             * This creates an instance that's a reference to the passed in
+             * cursor rather than copying it.  Side-effects to this object
+             * will affect the original cursor.
+             */
+            static ByteCursor Wrap(aws_byte_cursor *cursor) noexcept;
+
+            // APIs
+            void Advance(size_t len) noexcept;
+
+            // Accessors
+            aws_byte_cursor *GetImpl() noexcept { return m_cursorPtr; }
+            const aws_byte_cursor *GetImpl() const noexcept { return m_cursorPtr; }
+
+            const uint8_t *GetPtr() const noexcept { return m_cursorPtr->ptr; }
+            size_t GetLen() const noexcept { return m_cursorPtr->len; }
+
+          private:
+            aws_byte_cursor m_cursor;
+            aws_byte_cursor *m_cursorPtr;
+        };
+
+        class ByteBuf
+        {
+          public:
+            // Initialization that cannot fail
+            ByteBuf() noexcept;
+            ByteBuf(ByteBuf &&rhs) noexcept;
+            ByteBuf(const uint8_t *array, size_t capacity, size_t len) noexcept;
+
+            ByteBuf &operator=(ByteBuf &&buffer) noexcept;
+
+            ~ByteBuf();
+
+            /**
+             * This creates an instance that is a reference to the passed in
+             * buffer rather than copying it.  Side-effects to this object
+             * will affect the original c buffer.
+             */
+            static ByteBuf Wrap(aws_byte_buf *buffer) noexcept;
+
+            // Initialization that can fail
+            static AwsCrtResult<ByteBuf> Init(const ByteBuf &buffer) noexcept;
+            static AwsCrtResult<ByteBuf> Init(Allocator *alloc, size_t capacity) noexcept;
```

Comment:
I will say that having an init function versus a state variable has one single advantage; and that is it can be marked with `[[nodiscard]]` in the future. That being said, I still so much dislike the ctor + init pattern.

So, what are your thoughts about having an (optional?) error_callback function parameter to the ctors overloads that can fail?

Reply:
I don't see how failing assignment could be handled with that approach.  Having a callback that gets invoked in the middle of a construction failure (that still needs to be reacted to afterwards when your object is garbage) seems counter-intuitive.

I'm hoping the same error-handling pattern can be used not just for construction, but for all failable operations.  The fact that it's a little awkward in the construction case is ok in my mind currently.

Reply:
>I don't see how failing assignment could be handled with that approach.

I don't see how the current approach handles failing assignment.

Reply:
Can't fail what you don't allow =p